{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bellfort Sequence Parser"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import tkinter as tk\n",
    "from tkinter import ttk\n",
    "import tkinter.font as tkf\n",
    "from tkinter import messagebox\n",
    "from tkinter import filedialog\n",
    "import threading\n",
    "import time\n",
    "import os\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Helper Functions\n",
    "### Reverse Complement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def reverseComplement(sequence):\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C', 'N': 'N'}\n",
    "    rc_sequence=''\n",
    "    for s in sequence:\n",
    "        rc_sequence = complement[s] + rc_sequence\n",
    "    return rc_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FASTQ File Browse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def buttonBrowseFASTQ():\n",
    "    global filenameFASTQ, indicator_preprocess\n",
    "    \n",
    "    try:\n",
    "        filenameFASTQ = filedialog.askopenfilename(filetypes=(('FASTQ files', '*.fastq'), \n",
    "                                                              ('All files', '*.*')))\n",
    "        text_fileFASTQ.delete('1.0', tk.END)\n",
    "        text_fileFASTQ.insert('1.0', filenameFASTQ.split('/')[-1])\n",
    "        \n",
    "        # Reset the progress bar///////////////\n",
    "        progressbar['value'] = 0\n",
    "        progressbar_loadFASTQ['value'] = 0\n",
    "        \n",
    "        # Reset the percentage\n",
    "        text_percentage.delete('1.0', tk.END)\n",
    "        text_percentage.insert('1.0', str('0%'))\n",
    "        \n",
    "        indicator_preprocess = 0\n",
    "    except:\n",
    "        filenameFASTQ = ''   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FASTQ File Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def loadFASTQ():\n",
    "    global reads\n",
    "    \n",
    "    start_time = time.time()    \n",
    "    \n",
    "    f = open(filenameFASTQ)\n",
    "\n",
    "    reads = []\n",
    "\n",
    "    try:\n",
    "        while 1:\n",
    "            name = f.readline().rstrip()\n",
    "            sequence = f.readline().rstrip()\n",
    "            f.readline()\n",
    "            quality = f.readline().rstrip()\n",
    "\n",
    "            if len(name) == 0:\n",
    "                break\n",
    "\n",
    "            union = name, sequence\n",
    "\n",
    "            reads.append(union)           \n",
    "\n",
    "        end_time = time.time()\n",
    "        delta_time = end_time - start_time\n",
    "\n",
    "        text_time.delete('1.0', tk.END)\n",
    "        text_time.insert('1.0', str(delta_time))  \n",
    "\n",
    "        text_readNum.delete('1.0', tk.END)\n",
    "        text_readNum.insert('1.0', str(len(reads)))  \n",
    "\n",
    "    except:\n",
    "        messagebox.showwarning(\"File Loading Failed\", \n",
    "                               \"Sorry, file loading failed! Please check the file format.\")\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def start_loadFASTQ_thread(event):\n",
    "    global loadFASTQ_thread\n",
    "    \n",
    "    if filenameFASTQ != '':\n",
    "        loadFASTQ_thread = threading.Thread(target=loadFASTQ)\n",
    "        loadFASTQ_thread.daemon = True\n",
    "\n",
    "        progressbar_loadFASTQ.start(10)\n",
    "        loadFASTQ_thread.start()\n",
    "        root.after(20, check_loadFASTQ_thread)\n",
    "    else:\n",
    "        messagebox.showwarning(\"No File\", \n",
    "                               \"Sorry, no file loaded! Please choose FASTQ file first.\")\n",
    "\n",
    "def check_loadFASTQ_thread():\n",
    "    if loadFASTQ_thread.is_alive():\n",
    "        progressbar_loadFASTQ.start(10)\n",
    "        root.after(20, check_loadFASTQ_thread)\n",
    "    else:\n",
    "        progressbar_loadFASTQ.stop()\n",
    "        progressbar_loadFASTQ['value']=100\n",
    "        messagebox.showinfo(\"FASTQ File Loaded\", \"FASTQ file successfully loaded!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Divide FASTQ File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def divideFASTQ():    \n",
    "    start_time = time.time() \n",
    "\n",
    "    gotten = text_readNumDivided.get('1.0', tk.END)\n",
    "    readNumDivided = int(gotten.rstrip())\n",
    "\n",
    "    if os.path.exists(filenameFASTQ+'.folder'):           \n",
    "        # Remove the folder previously made:\n",
    "        shutil.rmtree(filenameFASTQ+'.folder')\n",
    "\n",
    "    # Make a new one:\n",
    "    os.makedirs(filenameFASTQ+'.folder')    \n",
    "\n",
    "    line_num = 0\n",
    "    file_no = 1\n",
    "\n",
    "    f_input = open(filenameFASTQ)        \n",
    "    f_output = open(filenameFASTQ+'.folder/' + filenameFASTQ.split('/')[-1] + '__Slice_No_' + str(file_no) + '.fastq', 'w') \n",
    "\n",
    "    while 1:\n",
    "        # Input ///////////////////////////////////\n",
    "        name = f_input.readline()\n",
    "        sequence = f_input.readline()\n",
    "        f_input.readline()\n",
    "        quality = f_input.readline()\n",
    "\n",
    "        if len(name) == 0:\n",
    "            break            \n",
    "\n",
    "        # Output ////////////////////////////////////\n",
    "\n",
    "        f_output.write(name)\n",
    "        f_output.write(sequence)\n",
    "        f_output.write('+\\n')\n",
    "        f_output.write(quality)      \n",
    "\n",
    "        line_num += 1\n",
    "\n",
    "        if line_num == readNumDivided:                \n",
    "            f_output.close() \n",
    "            file_no += 1\n",
    "            f_output = open(filenameFASTQ+'.folder/' + filenameFASTQ.split('/')[-1] + '__Slice_No_' + str(file_no) + '.fastq', 'w')\n",
    "            line_num = 0                \n",
    "\n",
    "    end_time = time.time()\n",
    "    delta_time = end_time - start_time\n",
    "\n",
    "    text_time.delete('1.0', tk.END)\n",
    "    text_time.insert('1.0', str(delta_time))  \n",
    "\n",
    "    f_input.close()\n",
    "    f_output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def start_divideFASTQ_thread(event):\n",
    "    global divideFASTQ_thread\n",
    "    \n",
    "    if filenameFASTQ != '':\n",
    "        divideFASTQ_thread = threading.Thread(target=divideFASTQ)\n",
    "        divideFASTQ_thread.daemon = True\n",
    "\n",
    "        progressbar_loadFASTQ.start(10)\n",
    "        divideFASTQ_thread.start()\n",
    "        root.after(20, check_divideFASTQ_thread)\n",
    "    else:\n",
    "        messagebox.showwarning(\"No File\", \n",
    "                               \"Sorry, no file loaded! Please choose FASTQ file first.\")\n",
    "\n",
    "def check_divideFASTQ_thread():\n",
    "    if divideFASTQ_thread.is_alive():\n",
    "        progressbar_loadFASTQ.start(10)\n",
    "        root.after(20, check_divideFASTQ_thread)\n",
    "    else:\n",
    "        progressbar_loadFASTQ.stop()\n",
    "        progressbar_loadFASTQ['value']=100\n",
    "        messagebox.showinfo(\"FASTQ File Divided\", \"FASTQ file has been successfully divided!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def preprocessFASTQ():\n",
    "    global reads, indicator_preprocess, kmer_dict_reads\n",
    "    \n",
    "    try:\n",
    "        num = len(reads)   \n",
    "        indicator_preprocess = 0\n",
    "        gain = 50/num\n",
    "\n",
    "        gotten = text_sequence_len.get('1.0', tk.END)\n",
    "        k = int(gotten.rstrip())\n",
    "        \n",
    "        if k > len(reads[0][1]):\n",
    "            messagebox.showwarning(\"Target Sequence Length Error\", \n",
    "                                   \"Sorry, the target sequence length is more than read length. Please check.\")\n",
    "        elif k < 3:\n",
    "            messagebox.showwarning(\"Sequence Too Short\", \n",
    "                                   \"Sorry, the target sequence length is too short which will make the program running slowly. Please check.\")\n",
    "        elif filenameSequences == '':\n",
    "            messagebox.showwarning(\"No Sequences Loaded\", \n",
    "                                   \"Sorry, no sequences loaded! Please load sequences first.\")\n",
    "        else:\n",
    "            kmer_dict_reads = {}\n",
    "\n",
    "            start_time = time.time()\n",
    "\n",
    "            for read in reads:\n",
    "                for i in range(len(read[1])-k+1):\n",
    "                    kmer_dict_reads[read[1][i:i+k]] = set()\n",
    "                indicator_preprocess += gain \n",
    "\n",
    "            for read in reads:\n",
    "                for i in range(len(read[1])-k+1):\n",
    "                    kmer_dict_reads[read[1][i:i+k]].add(read)\n",
    "                indicator_preprocess += gain\n",
    "                \n",
    "            indicator_progress = 100\n",
    "            \n",
    "            # Add MatchAll Here ///////////////////////////////////////////////////////\n",
    "            matchAll()\n",
    "\n",
    "            end_time = time.time()\n",
    "            delta_time = end_time - start_time\n",
    "\n",
    "            text_time.delete('1.0', tk.END)\n",
    "            text_time.insert('1.0', str(delta_time)) \n",
    "            \n",
    "            messagebox.showinfo(\"Preprocess FASTQ & Count Matched Sequences Completed\", \n",
    "                                \"Current FASTQ preprocess & matched sequence counts successfully completed!\")\n",
    "\n",
    "    except NameError:\n",
    "        messagebox.showwarning(\"No FASTQ File Loaded\", \n",
    "                               \"Sorry, no loaded FASTQ file found! Please load FASTQ file first.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def start_preprocess_thread(event):\n",
    "    global preprocess_thread, indicator_preprocess\n",
    "    preprocess_thread = threading.Thread(target=preprocessFASTQ)\n",
    "    preprocess_thread.daemon = True\n",
    "    \n",
    "    progressbar['value'] = indicator_preprocess\n",
    "    text_percentage.delete('1.0', tk.END)\n",
    "    text_percentage.insert('1.0', str(int(indicator_preprocess))+'%')\n",
    "    \n",
    "    preprocess_thread.start()\n",
    "    root.after(20, check_preprocess_thread)\n",
    "\n",
    "def check_preprocess_thread():\n",
    "    if preprocess_thread.is_alive():\n",
    "        progressbar['value'] = indicator_preprocess\n",
    "        text_percentage.delete('1.0', tk.END)\n",
    "        text_percentage.insert('1.0', str(int(indicator_preprocess))+'%')\n",
    "        \n",
    "        root.after(20, check_preprocess_thread)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Match All"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def matchAll():\n",
    "    global  kmer_dict_reads, indicator_matchAll, df\n",
    "    \n",
    "    try:\n",
    "        len(kmer_dict_reads)    \n",
    "        num = len(df)\n",
    "        \n",
    "        if num == 0:\n",
    "            messagebox.showwarning(\"No Sequences Loaded\", \n",
    "                                   \"Sorry, no sequences loaded! Please load sequences first.\")\n",
    "        else:    \n",
    "            indicator_matchAll = 0\n",
    "            gain = 1000000/num\n",
    "\n",
    "            start_time = time.time()\n",
    "\n",
    "            arr = np.array(df)\n",
    "\n",
    "            for i in range(len(arr)):\n",
    "                key1 = arr[i,2]\n",
    "                key2 = reverseComplement(key1)\n",
    "                \n",
    "                try:\n",
    "                    n1 = len(kmer_dict_reads[key1])\n",
    "                except KeyError:\n",
    "                    n1 = 0\n",
    "                    \n",
    "                try:\n",
    "                    n2 = len(kmer_dict_reads[key2])\n",
    "                except KeyError:\n",
    "                    n2 = 0\n",
    "                    \n",
    "                arr[i, 4] += n1 + n2\n",
    "                arr[i, 5] += 1\n",
    "                \n",
    "                indicator_matchAll += gain\n",
    "\n",
    "            df = pd.DataFrame(arr, columns = ['gene_id', 'UID', 'seq', 'Reserved', 'Count', 'Tag'])\n",
    "            #df = df.set_index('UID', drop=False) \n",
    "\n",
    "            end_time = time.time()\n",
    "            delta_time = end_time - start_time\n",
    "\n",
    "            text_time.delete('1.0', tk.END)\n",
    "            text_time.insert('1.0', str(delta_time))           \n",
    "\n",
    "    except NameError:\n",
    "        messagebox.showwarning(\"No FASTQ Preprocessed or No Sequences Loaded\", \n",
    "                               \"Sorry, no FASTQ preprocess implemented or no sequences file loaded! Please preprocess FASTQ or load sequences first.\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def start_matchAll_thread(event):\n",
    "    global matchAll_thread, indicator_matchAll\n",
    "    matchAll_thread = threading.Thread(target=matchAll)\n",
    "    matchAll_thread.daemon = True\n",
    "    \n",
    "    progressbar['value'] = indicator_matchAll\n",
    "    \n",
    "    matchAll_thread.start()\n",
    "    root.after(20, check_matchAll_thread)\n",
    "\n",
    "def check_matchAll_thread():\n",
    "    if matchAll_thread.is_alive():\n",
    "        progressbar['value'] = indicator_matchAll\n",
    "        \n",
    "        root.after(20, check_matchAll_thread)\n",
    "    else:\n",
    "        messagebox.showinfo(\"Matching Completed\", \n",
    "                                \"Counting of sequences matched successfully completed!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Match Single"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def buttonMatch():\n",
    "    gotten = text_sequence.get('1.0', tk.END)\n",
    "    p1 = gotten.rstrip()    \n",
    "    p2 = reverseComplement(p1)\n",
    "    \n",
    "    if p1 == '' or p2 == '':\n",
    "        messagebox.showwarning(\"No Sequence Found\", \n",
    "                               \"Sorry, no sequence found in the text blank above! Please check the sequence.\")\n",
    "    else:\n",
    "        try:\n",
    "            len(kmer_dict_reads)\n",
    "            try:\n",
    "                n1 = len(kmer_dict_reads[p1])\n",
    "            except KeyError:\n",
    "                n1 = 0\n",
    "            \n",
    "            try:\n",
    "                n2 = len(kmer_dict_reads[p2])\n",
    "            except KeyError:\n",
    "                n2 = 0\n",
    "                \n",
    "            count = n1 + n2\n",
    "                \n",
    "            text_count.delete('1.0', tk.END)\n",
    "            text_count.insert('1.0', str(count))\n",
    "            \n",
    "        except NameError:\n",
    "            messagebox.showwarning(\"No FASTQ Preprocessed\", \n",
    "                                   \"Sorry, no FASTQ preprocess implemented! Please preprocess FASTQ first.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File of Target Sequence Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def loadSequences():\n",
    "    global filenameSequences, df, recordNum\n",
    "    \n",
    "    progressbar_loadSequences['value'] = 0\n",
    "    try:\n",
    "        filenameSequences = filedialog.askopenfilename(filetypes=(('Comma-Separated (CSV) text file', '*.csv'), ('All files', '*.*')))\n",
    "        text_fileSequences.delete('1.0', tk.END)\n",
    "        text_fileSequences.insert('1.0', filenameSequences.split('/')[-1])\n",
    "    except:\n",
    "        filenameSequences = ''    \n",
    "   \n",
    "    if filenameSequences == '':\n",
    "        messagebox.showwarning(\"No File\", \"Sorry, no file chosen! Please choose file of sequences first.\")\n",
    "    else:        \n",
    "        try:\n",
    "            start_time = time.time()\n",
    "            \n",
    "            df = pd.read_csv(filenameSequences)\n",
    "            df['count'] = 0\n",
    "            df['tag'] = 0\n",
    "            #df = df.set_index('UID', drop=False)  \n",
    "            \n",
    "            recordNum = len(df)\n",
    "            \n",
    "            progressbar_loadSequences['value'] = 100\n",
    "            \n",
    "            end_time = time.time()\n",
    "            delta_time = end_time - start_time\n",
    "                       \n",
    "            text_time.delete('1.0', tk.END)\n",
    "            text_time.insert('1.0', str(delta_time))\n",
    "            \n",
    "            text_recordNum.delete('1.0', tk.END)\n",
    "            text_recordNum.insert('1.0', str(recordNum))\n",
    "            \n",
    "            messagebox.showinfo(\"File of Sequences Loaded\", \"File of sequences successfully loaded!\")        \n",
    "        except:\n",
    "            messagebox.showwarning(\"File Loading Failed\", \"Sorry, file loading failed! Please check the file format.\")    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Half-matched Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def buttonLoadHalfMatchedSequences():\n",
    "    global df, recordNum, filenameSequences\n",
    "    \n",
    "    progressbar_loadSequences['value'] = 0\n",
    "    try:\n",
    "        filenameSequences = filedialog.askopenfilename(filetypes=(('Comma-Separated (CSV) text file', '*.csv'), ('All files', '*.*')))\n",
    "        text_fileSequences.delete('1.0', tk.END)\n",
    "        text_fileSequences.insert('1.0', filenameSequences.split('/')[-1])\n",
    "    except:\n",
    "        filenameSequences = ''    \n",
    "        \n",
    "    if filenameSequences == '':\n",
    "        messagebox.showwarning(\"No File\", \"Sorry, no file chosen! Please choose file of sequences first.\")\n",
    "    else:        \n",
    "        try:\n",
    "            start_time = time.time()\n",
    "            \n",
    "            df = pd.read_csv(filenameSequences)   \n",
    "            df = df.set_index('Unnamed: 0', drop=True)  \n",
    "            \n",
    "            recordNum = len(df)\n",
    "            \n",
    "            progressbar_loadSequences['value'] = 100\n",
    "            \n",
    "            end_time = time.time()\n",
    "            delta_time = end_time - start_time\n",
    "                       \n",
    "            text_time.delete('1.0', tk.END)\n",
    "            text_time.insert('1.0', str(delta_time))\n",
    "            \n",
    "            text_recordNum.delete('1.0', tk.END)\n",
    "            text_recordNum.insert('1.0', str(recordNum))\n",
    "            \n",
    "            messagebox.showinfo(\"File of Half Matched Sequences Loaded\", \"File of half matched sequences successfully loaded!\")        \n",
    "        except:\n",
    "            messagebox.showwarning(\"File Loading Failed\", \"Sorry, file loading failed! Please check the file format.\")    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Table Events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def OnDoubleClick(event):\n",
    "    item = table.selection()[0]\n",
    "    value = table.item(item, 'values')\n",
    "    geneID = value[0]\n",
    "    uid = value[1]\n",
    "    sequence = value[2]\n",
    "    rc_sequence = reverseComplement(sequence)\n",
    "    \n",
    "    text_geneID.delete('1.0', tk.END)\n",
    "    text_geneID.insert('1.0', str(geneID))\n",
    "    \n",
    "    text_uid.delete('1.0', tk.END)\n",
    "    text_uid.insert('1.0', str(uid))\n",
    "    \n",
    "    text_sequence.delete('1.0', tk.END)\n",
    "    text_sequence.insert('1.0', str(sequence))\n",
    "    \n",
    "    text_rc_sequence.delete('1.0', tk.END)\n",
    "    text_rc_sequence.insert('1.0', str(rc_sequence))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def sortby(tree, col, descending):\n",
    "    \"\"\"sort tree contents when a column header is clicked on\"\"\"\n",
    "    # grab values to sort\n",
    "    data = [(tree.set(child, col), child) for child in tree.get_children('')]\n",
    "    # if the data to be sorted is numeric change to float\n",
    "    #data =  change_numeric(data)\n",
    "    # now sort the data in place\n",
    "    data.sort(reverse=descending)\n",
    "    for ix, item in enumerate(data):\n",
    "        tree.move(item[1], '', ix)\n",
    "    # switch the heading so it will sort in the opposite direction\n",
    "    tree.heading(col, command=lambda col=col: sortby(tree, col, int(not descending)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def display_in_table():\n",
    "    try:\n",
    "        for a in df.index:\n",
    "            row = df.ix[a]\n",
    "            table.insert(\"\", \"end\", \"\", values=tuple(row)) \n",
    "    except NameError:\n",
    "        messagebox.showwarning(\"No Sequences to be Displayed\", \n",
    "                               \"Sorry, there's no loaded sequences to be displayed! Please load sequence file first.\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Other Button Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def clear():\n",
    "    for i in table.get_children():\n",
    "        table.delete(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def browse():\n",
    "    start_time = time.time()\n",
    "    clear()\n",
    "    display_in_table()\n",
    "    delta_time = time.time() - start_time\n",
    "    \n",
    "    text_time.delete('1.0', tk.END)\n",
    "    text_time.insert('1.0', str(delta_time))           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def buttonExport():   \n",
    "    if filenameSequences == '' or (filenameFASTQ == '' and len(filenameFASTQs) == 0):\n",
    "        messagebox.showwarning(\"No File Loaded\", \n",
    "                               \"Sorry, no file loaded! Please choose sequence file and FASTQ file first.\")\n",
    "    else:\n",
    "        try:\n",
    "            len(df)\n",
    "            len(reads)\n",
    "            directory = filedialog.askdirectory()\n",
    "            df.to_csv(directory + '/Counts of ' + filenameSequences.split('/')[-1] + '__matchedWith__' + filenameFASTQs[0].split('/')[-1] + '.csv')\n",
    "            messagebox.showinfo(\"File Exported\", \"File of counted sequences successfully exported!\")        \n",
    "        except NameError:\n",
    "            messagebox.showwarning(\"Error: No Counted DataFrame Generated\", \n",
    "                               \"Sorry, no effective counted DataFrame generated! Please check the previous workflow.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def buttonAbout():\n",
    "    about_root=tk.Tk()\n",
    "    \n",
    "    w = 380 # width for the Tk root\n",
    "    h = 310 # height for the Tk root\n",
    "\n",
    "    # get screen width and height\n",
    "    ws = about_root.winfo_screenwidth() # width of the screen\n",
    "    hs = about_root.winfo_screenheight() # height of the screen\n",
    "\n",
    "    # calculate x and y coordinates for the Tk root window\n",
    "    x = (ws/2) - (w/2)\n",
    "    y = (hs/2) - (h/2)\n",
    "\n",
    "    # set the dimensions of the screen \n",
    "    # and where it is placed\n",
    "    about_root.geometry('%dx%d+%d+%d' % (w, h, x, y))\n",
    "    about_root.title('About Bellfort Sequence Parser')  \n",
    "    about_root.iconbitmap('dna.ico')\n",
    "\n",
    "    label_author=tk.Label(about_root,text='Bellfort Sequence Parser Version 2.0', font=('tahoma', 9))\n",
    "    label_author.place(x=90,y=30)\n",
    "\n",
    "    label_author=tk.Label(about_root,text='Copyright (C) 2016', font=('tahoma', 9))\n",
    "    label_author.place(x=125,y=60)\n",
    "    \n",
    "    label_author=tk.Label(about_root,text='Chen Lab', font=('tahoma', 9))\n",
    "    label_author.place(x=150,y=90)\n",
    "    \n",
    "    label_author=tk.Label(about_root,text='Human Genome Sequencing Center', font=('tahoma', 9))\n",
    "    label_author.place(x=80,y=120)\n",
    "    \n",
    "    label_author=tk.Label(about_root,text='Department of Molecular and Human Genetics', font=('tahoma', 9))\n",
    "    label_author.place(x=50,y=150)\n",
    "    \n",
    "    label_author=tk.Label(about_root,text='Baylor College of Medicine', font=('tahoma', 9))\n",
    "    label_author.place(x=110,y=180)\n",
    "   \n",
    "\n",
    "    button_okay=ttk.Button(about_root, width=15, text='OK', command=about_root.destroy)\n",
    "    button_okay.place(x=130, y=235)\n",
    "\n",
    "    about_root.mainloop()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Batch Mode\n",
    "#### FASTQ Files Loaded Button"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def buttonBrowseFASTQs():\n",
    "    global filenameFASTQs\n",
    "    \n",
    "    try:\n",
    "        filenameFASTQs = filedialog.askopenfilenames(filetypes=(('FASTQ files', '*.fastq'), \n",
    "                                                              ('All files', '*.*')))\n",
    "        text_fileFASTQs.delete('1.0', tk.END)\n",
    "        text_fileFASTQs.insert('1.0', filenameFASTQs[0].split('/')[-1].split('__')[0])\n",
    "        \n",
    "        text_fileNum.delete('1.0', tk.END)\n",
    "        text_fileNum.insert('1.0', str(len(filenameFASTQs)))\n",
    "        \n",
    "        # Reset the progress bar///////////////\n",
    "        progressbar['value'] = 0\n",
    "        progressbar_loadFASTQ['value'] = 0\n",
    "        \n",
    "        # Reset the percentage\n",
    "        text_percentage.delete('1.0', tk.END)\n",
    "        text_percentage.insert('1.0', str('0%'))\n",
    "        \n",
    "        indicator_preprocess = 0\n",
    "    except:\n",
    "        filenameFASTQs = ''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Batch Process Button Series"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def loadFASTQ_batch(filenameFASTQ):\n",
    "    global reads\n",
    "    \n",
    "    start_time = time.time() \n",
    "    \n",
    "    f = open(filenameFASTQ)\n",
    "\n",
    "    reads = []\n",
    "\n",
    "    #try:\n",
    "    while 1:\n",
    "        name = f.readline().rstrip()\n",
    "        sequence = f.readline().rstrip()\n",
    "        f.readline()\n",
    "        quality = f.readline().rstrip()\n",
    "\n",
    "        if len(name) == 0:\n",
    "            break\n",
    "\n",
    "        union = name, sequence, quality\n",
    "\n",
    "        reads.append(union)\n",
    "\n",
    "    f.close()\n",
    "\n",
    "    end_time = time.time()\n",
    "    delta_time = end_time - start_time\n",
    "\n",
    "    text_time.delete('1.0', tk.END)\n",
    "    text_time.insert('1.0', str(delta_time))  \n",
    "\n",
    "    text_readNum.delete('1.0', tk.END)\n",
    "    text_readNum.insert('1.0', str(len(reads)))  \n",
    "\n",
    "    '''\n",
    "    except:\n",
    "        messagebox.showwarning(\"File Loading Failed\", \n",
    "                               \"Sorry, file loading failed! Please check the file format.\")    '''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def preprocessFASTQ_batch():\n",
    "    global reads, kmer_dict_reads, indicator_batch, gain_file\n",
    "    \n",
    "    try:  \n",
    "        gotten = text_sequence_len.get('1.0', tk.END)\n",
    "        k = int(gotten.rstrip())\n",
    "        \n",
    "        if k > len(reads[0][1]):\n",
    "            messagebox.showwarning(\"Target Sequence Length Error\", \n",
    "                                   \"Sorry, the target sequence length is more than read length. Please check.\")\n",
    "        elif k < 3:\n",
    "            messagebox.showwarning(\"Sequence Too Short\", \n",
    "                                   \"Sorry, the target sequence length is too short which will make the program running slowly. Please check.\")\n",
    "        elif filenameSequences == '':\n",
    "            messagebox.showwarning(\"No Sequences Loaded\", \n",
    "                                   \"Sorry, no sequences loaded! Please load sequences first.\")\n",
    "        else:\n",
    "            kmer_dict_reads = {}\n",
    "                        \n",
    "            gain = gain_file/(len(reads)*2)\n",
    "\n",
    "            for read in reads:\n",
    "                for i in range(len(read[1])-k+1):\n",
    "                    kmer_dict_reads[read[1][i:i+k]] = set()\n",
    "                    \n",
    "                indicator_batch += gain \n",
    "                \n",
    "            for read in reads:\n",
    "                for i in range(len(read[1])-k+1):\n",
    "                    kmer_dict_reads[read[1][i:i+k]].add(read)\n",
    "                    \n",
    "                indicator_batch += gain \n",
    "\n",
    "    except NameError:\n",
    "        messagebox.showwarning(\"No FASTQ File Loaded\", \n",
    "                               \"Sorry, no loaded FASTQ file found! Please load FASTQ file first.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def matchAll_batch():\n",
    "    global  kmer_dict_reads, df, indicator_batch, gain, reads_matched\n",
    "    \n",
    "    try:       \n",
    "        arr = np.array(df)\n",
    "        \n",
    "        reads_matched = set()\n",
    "\n",
    "        for i in range(len(arr)):\n",
    "            key1 = arr[i,2]\n",
    "            key2 = reverseComplement(key1)\n",
    "\n",
    "            try:\n",
    "                n1 = len(kmer_dict_reads[key1])\n",
    "            except KeyError:\n",
    "                n1 = 0\n",
    "\n",
    "            try:\n",
    "                n2 = len(kmer_dict_reads[key2])\n",
    "            except KeyError:\n",
    "                n2 = 0\n",
    "                \n",
    "            arr[i, 4] += n1 + n2\n",
    "            arr[i, 5] += 1\n",
    "            \n",
    "            if n1 != 0:\n",
    "                for read in kmer_dict_reads[key1]:\n",
    "                    start_point = read[1].find(key1)\n",
    "                    reads_matched.add((read[0], key1, read[2][start_point : start_point+len(key1)]))\n",
    "                    \n",
    "            if n2 != 0:\n",
    "                for read in kmer_dict_reads[key2]:\n",
    "                    start_point = read[1].find(key2)\n",
    "                    reads_matched.add((read[0], key2, read[2][start_point : start_point+len(key2)]))\n",
    "            \n",
    "\n",
    "\n",
    "        df = pd.DataFrame(arr, columns = ['gene_id', 'UID', 'seq', 'Reserved', 'Count', 'Tag'])\n",
    "        #df = df.set_index('UID', drop=False) \n",
    "\n",
    "    except NameError:\n",
    "        messagebox.showwarning(\"No FASTQ Preprocessed or No Sequences Loaded\", \n",
    "                               \"Sorry, no FASTQ preprocess implemented or no sequences file loaded! Please preprocess FASTQ or load sequences first.\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def autoExport():   \n",
    "    if filenameSequences == '' or (filenameFASTQ == '' and len(filenameFASTQs) == 0):\n",
    "        messagebox.showwarning(\"No File Loaded\", \n",
    "                               \"Sorry, no file loaded! Please choose sequence file and FASTQ file first.\")\n",
    "    else:\n",
    "        try:\n",
    "            len(df)\n",
    "            len(reads)\n",
    "            \n",
    "            df.to_csv(filenameSequences + '__matchedWith__' + filenameFASTQs[0].split('/')[-1].split('__')[0] + '.csv')\n",
    "            \n",
    "        except NameError:\n",
    "            messagebox.showwarning(\"Error: No Counted DataFrame Generated\", \n",
    "                               \"Sorry, no effective counted DataFrame generated! Please check the previous workflow.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def batchProcess():\n",
    "    global filenameFASTQs, indicator_batch, gain_file, kmer_dict_reads, reads_matched\n",
    "    \n",
    "    start_time = time.time() \n",
    "    \n",
    "    if filenameFASTQs == '':\n",
    "        messagebox.showwarning('No FASTQ Chosen', \n",
    "                               'Sorry, no FASTQ file chosen! Please browse and choose FASTQ file first.')   \n",
    "        \n",
    "    elif len(df) == 0:\n",
    "        messagebox.showwarning(\"No Sequences Loaded\", \n",
    "                                   \"Sorry, no sequences loaded! Please load sequences first.\")\n",
    "    else:\n",
    "        indicator_batch = 0\n",
    "        gain_file = 100/len(filenameFASTQs)\n",
    "        \n",
    "        # ///////////////// Main Stream of Batch Match ////////////////////////////////////////////////////\n",
    "        \n",
    "        f_trimmed = open(filenameFASTQs[0].split('__')[0] + '__trimmedBy__' + filenameSequences.split('/')[-1] + '.fastq', 'w')\n",
    "        \n",
    "        reads_matched = set()\n",
    "        \n",
    "        for filenameFASTQ in filenameFASTQs:\n",
    "            loadFASTQ_batch(filenameFASTQ)\n",
    "            preprocessFASTQ_batch()\n",
    "            matchAll_batch()\n",
    "            \n",
    "            for read_matched in reads_matched:\n",
    "                f_trimmed.write(read_matched[0]+'\\n')\n",
    "                f_trimmed.write(read_matched[1]+'\\n')\n",
    "                f_trimmed.write('+\\n')\n",
    "                f_trimmed.write(read_matched[2]+'\\n')\n",
    "                \n",
    "        f_trimmed.close()\n",
    "        \n",
    "        autoExport()  \n",
    "        \n",
    "        # Attention! Memory Recollected !!!!!!!!!!\n",
    "        kmer_dict_reads = {}\n",
    "        \n",
    "        indicator_batch = 100\n",
    "            \n",
    "        messagebox.showinfo('Matching Completed', 'Tada! Counting of sequences matched successfully completed!')  \n",
    "        \n",
    "    delta_time = time.time() - start_time\n",
    "    \n",
    "    \n",
    "    \n",
    "    text_time.delete('1.0', tk.END)\n",
    "    text_time.insert('1.0', str(delta_time))   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def start_batch_thread(event):\n",
    "    global batch_thread, indicator_batch\n",
    "    batch_thread = threading.Thread(target=batchProcess)\n",
    "    batch_thread.daemon = True\n",
    "    \n",
    "    progressbar_batch['value'] = indicator_batch\n",
    "    text_percentage_batch.delete('1.0', tk.END)\n",
    "    text_percentage_batch.insert('1.0', str(int(indicator_batch))+'%')\n",
    "    \n",
    "    batch_thread.start()\n",
    "    root.after(20, check_batch_thread)\n",
    "\n",
    "def check_batch_thread():\n",
    "    if batch_thread.is_alive():\n",
    "        progressbar_batch['value'] = indicator_batch\n",
    "        text_percentage_batch.delete('1.0', tk.END)\n",
    "        text_percentage_batch.insert('1.0', str(int(indicator_batch))+'%')\n",
    "        \n",
    "        root.after(20, check_batch_thread)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main Flow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "headers = ['gene_id', 'UID', 'seq', 'Reserved', 'count', 'tag']\n",
    "header_widths = [280, 150, 350, 100, 80, 100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "root = tk.Tk()\n",
    "\n",
    "indicator_preprocess = 0\n",
    "indicator_loadSequences = 0\n",
    "indicator_matchAll = 0\n",
    "indicator_batch = 0\n",
    "filenameSequences = ''\n",
    "filenameFASTQ = ''\n",
    "recordNum = 0\n",
    "count = 0\n",
    "df = pd.DataFrame([])\n",
    "\n",
    "root.geometry(\"{0}x{1}+0+0\".format(root.winfo_screenwidth(), root.winfo_screenheight()))\n",
    "#root.attributes('-fullscreen', True)\n",
    "root.title('Bellfort Sequence Parser')\n",
    "root.iconbitmap('dna.ico')\n",
    "\n",
    "\n",
    "# Multicolumn Listbox/////////////////////////////////////////////////////////////////////////////\n",
    "table = ttk.Treeview(height=\"20\", columns=headers, selectmode=\"extended\")\n",
    "table.pack(padx=10, pady=20, ipadx=1200, ipady=100)\n",
    "\n",
    "i = 1\n",
    "for header in headers:\n",
    "    table.heading('#'+str(i), text=header.title(), anchor=tk.W, command=lambda c=header: sortby(table, c, 0))\n",
    "    table.column('#'+str(i), stretch=tk.NO, minwidth=0, width=tkf.Font().measure(header.title())+header_widths[i-1]) \n",
    "    i+=1    \n",
    "table.column('#0', stretch=tk.NO, minwidth=0, width=0)\n",
    "\n",
    "table.bind(\"<Double-1>\", OnDoubleClick)\n",
    "#///////////////////////////////////////////////////////////////////////////////////////////\n",
    "\n",
    "# Scrollbar////////////////////////////////////////////////////////////////////////////////////////\n",
    "vsb = ttk.Scrollbar(table, orient=\"vertical\",  command = table.yview)\n",
    "hsb = ttk.Scrollbar(table, orient=\"horizontal\", command = table.xview)\n",
    "## Link scrollbars activation to top-level object\n",
    "table.configure(yscrollcommand=vsb.set, xscrollcommand=hsb.set)\n",
    "## Link scrollbar also to every columns\n",
    "map(lambda col: col.configure(yscrollcommand=vsb.set,xscrollcommand=hsb.set), table)\n",
    "vsb.pack(side = tk.RIGHT, fill = tk.Y)\n",
    "hsb.pack(side = tk.BOTTOM, fill = tk.X)        \n",
    "\n",
    "#//////////////////////////////////////////////////////////////////////////////////////////////\n",
    "y0 =310\n",
    "y1 = 350\n",
    "y2 = 420\n",
    "y3 = 460\n",
    "y4 = 520\n",
    "y5 = 555\n",
    "y6 = 595\n",
    "y7 = 645\n",
    "y8 = 695\n",
    "# Text /////////////////////////////////////////////////////////////////////////////////////\n",
    "text_recordNum=tk.Text(root, width=18, height=1, font=('tahoma', 9), bd=2, wrap='none')\n",
    "text_recordNum.place(x=840, y=y0)\n",
    "label_recordNum=tk.Label(root, text='records', font=('tahoma', 9))\n",
    "label_recordNum.place(x=1000,y=y0)\n",
    "\n",
    "text_fileSequences=tk.Text(root, width=55, height=1, font=('tahoma', 9), bd=2, wrap='none')\n",
    "text_fileSequences.place(x=60, y=y0)\n",
    "\n",
    "text_fileFASTQ=tk.Text(root, width=36, height=1, font=('tahoma', 9), bd=2, wrap='none')\n",
    "text_fileFASTQ.place(x=60, y=y4)\n",
    "\n",
    "text_fileFASTQs=tk.Text(root, width=36, height=1, font=('tahoma', 9), bd=2, wrap='none')\n",
    "text_fileFASTQs.place(x=60, y=y6+10)\n",
    "\n",
    "text_count=tk.Text(root, width=16, height=1, font=('tahoma', 9), bd=2)\n",
    "text_count.place(x=1000, y=y3)\n",
    "label_count=tk.Label(root, text='Count:', font=('tahoma', 9))\n",
    "label_count.place(x=940,y=y3)\n",
    "\n",
    "text_geneID=tk.Text(root, width=20, height=1, font=('tahoma', 9), bd=2)\n",
    "text_geneID.place(x=140, y=y2)\n",
    "label_geneID=tk.Label(root, text='Gene ID:', font=('tahoma', 9))\n",
    "label_geneID.place(x=60,y=y2)\n",
    "\n",
    "text_uid=tk.Text(root, width=20, height=1, font=('tahoma', 9), bd=2)\n",
    "text_uid.place(x=390, y=y2)\n",
    "label_uid=tk.Label(root, text='UID:', font=('tahoma', 9))\n",
    "label_uid.place(x=340,y=y2)\n",
    "\n",
    "text_sequence=tk.Text(root, width=38, height=1, font=('tahoma', 9), bd=2)\n",
    "text_sequence.place(x=680, y=y2)\n",
    "label_sequence=tk.Label(root, text='Sequence:', font=('tahoma', 9))\n",
    "label_sequence.place(x=600,y=y2)\n",
    "\n",
    "text_rc_sequence=tk.Text(root, width=38, height=1, font=('tahoma', 9), bd=2)\n",
    "text_rc_sequence.place(x=1000, y=y2)\n",
    "\n",
    "text_sequence_len=tk.Text(root, width=5, height=1, font=('tahoma', 9), bd=2)\n",
    "text_sequence_len.place(x=1260, y=y5)\n",
    "label_sequence_len=tk.Label(root, text='nts', font=('tahoma', 9))\n",
    "label_sequence_len.place(x=1315,y=y5)\n",
    "text_sequence_len.delete('1.0', tk.END)\n",
    "text_sequence_len.insert('1.0', str(20))\n",
    "\n",
    "text_readNumDivided=tk.Text(root, width=13, height=1, font=('tahoma', 9), bd=2, wrap='none')\n",
    "text_readNumDivided.place(x=335, y=y3+10)\n",
    "label_readNumDivided1=tk.Label(root, text='by every', font=('tahoma', 9))\n",
    "label_readNumDivided1.place(x=255,y=y3+10)\n",
    "label_readNumDivided2=tk.Label(root, text='reads', font=('tahoma', 9))\n",
    "label_readNumDivided2.place(x=460,y=y3+10)\n",
    "text_readNumDivided.delete('1.0', tk.END)\n",
    "text_readNumDivided.insert('1.0', str(250000))\n",
    "\n",
    "text_fileNum=tk.Text(root, width=12, height=1, font=('tahoma', 9), bd=2, wrap='none')\n",
    "text_fileNum.place(x=400, y=y6+10)\n",
    "label_fileNum=tk.Label(root, text='files', font=('tahoma', 9))\n",
    "label_fileNum.place(x=520,y=y6+10)\n",
    "\n",
    "text_readNum=tk.Text(root, width=22, height=1, font=('tahoma', 9), bd=2, wrap='none')\n",
    "text_readNum.place(x=700, y=y6+10)\n",
    "label_readNum=tk.Label(root, text='reads', font=('tahoma', 9))\n",
    "label_readNum.place(x=890,y=y6+10)\n",
    "\n",
    "text_time=tk.Text(root, width=15, height=1, font=('tahoma', 9), bd=2)\n",
    "text_time.place(x=115, y=y8)\n",
    "label_time=tk.Label(root, text='Time:', font=('tahoma', 9))\n",
    "label_time.place(x=60,y=y8)\n",
    "label_seconds=tk.Label(root, text='second(s)', font=('tahoma', 9))\n",
    "label_seconds.place(x=250,y=y8)\n",
    "\n",
    "text_percentage=tk.Text(root, width=8, height=1, font=('tahoma', 9), bg='gray95', bd=0)\n",
    "text_percentage.place(x=1260, y=y4)\n",
    "\n",
    "text_percentage_batch=tk.Text(root, width=8, height=1, font=('tahoma', 9), bg='gray95', bd=0)\n",
    "text_percentage_batch.place(x=1260, y=y7)\n",
    "\n",
    "# ProgressBar /////////////////////////////////////////////////////////////////////////////\n",
    "progressbar_loadSequences = ttk.Progressbar(root, length=200, maximum=100, mode='determinate')\n",
    "progressbar_loadSequences.place(x=530,y=y0)\n",
    "\n",
    "progressbar_loadFASTQ = ttk.Progressbar(root, length=300, mode='indeterminate')\n",
    "progressbar_loadFASTQ.place(x=400,y=y4)\n",
    "\n",
    "progressbar = ttk.Progressbar(root, length=460, maximum=100, mode='determinate')\n",
    "progressbar.place(x=760,y=y4)\n",
    "\n",
    "progressbar_batch = ttk.Progressbar(root, length=520, maximum=100, mode='determinate')\n",
    "progressbar_batch.place(x=700,y=y7)\n",
    "\n",
    "# Button /////////////////////////////////////////////////////////////////////////////////\n",
    "button_loadSequences = ttk.Button(root, text=\"Load sgRNA\", width=20, command=loadSequences)\n",
    "button_loadSequences.place(x=60, y=y1)\n",
    "\n",
    "button_loadHalfMatchedSequences = ttk.Button(root, text=\"Load Half Matched sgRNA\", width=30, command=buttonLoadHalfMatchedSequences)\n",
    "button_loadHalfMatchedSequences.place(x=265, y=y1)\n",
    "\n",
    "button_clear = ttk.Button(root, text=\"Clear\", width=20, command=clear)\n",
    "button_clear.place(x=1180, y=y1)\n",
    "\n",
    "button_refresh = ttk.Button(root, text=\"Browse\", width=20, command=browse)\n",
    "button_refresh.place(x=1180, y=y0)\n",
    "\n",
    "button_browseFASTQ = ttk.Button(root, text=\"Browse FASTQ...\", width=20, command=buttonBrowseFASTQ)\n",
    "button_browseFASTQ.place(x=60, y=y5)\n",
    "\n",
    "button_divideFASTQ = ttk.Button(root, text=\"Divide FASTQ\", width=20, command=lambda:start_divideFASTQ_thread(None))\n",
    "button_divideFASTQ.place(x=60, y=y3+10)\n",
    "\n",
    "button_loadFASTQ = ttk.Button(root, text=\"Load FASTQ\", width=20, command=lambda:start_loadFASTQ_thread(None))\n",
    "button_loadFASTQ.place(x=400, y=y5)\n",
    "\n",
    "button_preprocessFASTQ = ttk.Button(root, text=\"Preprocess FASTQ & Count All Matched Sequences\", width=55, command=lambda:start_preprocess_thread(None))\n",
    "button_preprocessFASTQ.place(x=760, y=y5)\n",
    "\n",
    "button_match = ttk.Button(root, text=\"Match\", width=20, command=buttonMatch)\n",
    "button_match.place(x=680, y=y3)\n",
    "\n",
    "button_matchAll = ttk.Button(root, text=\"Match All\", width=20, command=lambda:start_matchAll_thread(None))\n",
    "button_matchAll.place(x=1180, y=y3)\n",
    "\n",
    "button_about = ttk.Button(root, text=\"About\", width=20, command=buttonAbout)\n",
    "button_about.place(x=980, y=y8)\n",
    "\n",
    "button_export = ttk.Button(root, text=\"Export\", width=20, command=buttonExport)\n",
    "button_export.place(x=720, y=y8)\n",
    "\n",
    "button_exit = ttk.Button(root, text=\"Exit\", width=20, command=root.destroy)\n",
    "button_exit.place(x=1180, y=y8)\n",
    "\n",
    "button_browseFASTQs = ttk.Button(root, text=\"Browse FASTQs...\", width=25, command=buttonBrowseFASTQs)\n",
    "button_browseFASTQs.place(x=60, y=y7)\n",
    "\n",
    "button_batchProcess = ttk.Button(root, text=\"Batch Process\", width=30, command=lambda:start_batch_thread(None))\n",
    "button_batchProcess.place(x=400, y=y7)\n",
    "\n",
    "root.bind('<Return>', start_preprocess_thread)\n",
    "root.bind('<Return>', start_loadFASTQ_thread)\n",
    "root.bind('<Return>', start_divideFASTQ_thread)\n",
    "root.bind('<Return>', start_matchAll_thread)\n",
    "\n",
    "root.mainloop()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
